src/core/context/AppContextProvider.tsx

```diff
@@ -543,12 +573,32 @@ async function syncWithLocalStorage(
   );
 
   const existingConversation = allConversations[conversationID];
+
+  // Calculate tokens remaining - use from API response if available
+  let tokensRemaining: number;
+  if (apiResponse && apiResponse.usage && apiResponse.usage.total_tokens) {
+    tokensRemaining = Math.max(
+      0,
+      contextLength - apiResponse.usage.total_tokens,
+    );
+  } else {
+    //  Fallback to estimation if no token info available
+    const estimatedUsage =
+      //  todo - move this utility to the model config
+      modelConfig.id === 'deepseek-r1'
+        ? estimateTokenUsage(messages)
+        : //  this will be replaced by a gpt-specific utility
+          { totalTokens: 0 };
+    tokensRemaining = Math.max(0, contextLength - estimatedUsage.totalTokens);
+  }
```

Comment:
This will eventually become something like `modelConfig.contextLimitMonitor(messages)` and then the model config will specify which function to run.